File: src/Systems/C++/decorator.cpp
#include <iostream>
#include <memory>
#include <string>

class Component {
public:
    virtual ~Component() = default;
    [[nodiscard]] virtual std::string render() const = 0;
};

class PlainMessage final : public Component {
public:
    [[nodiscard]] std::string render() const override { return "alert"; }
};

class ComponentDecorator : public Component {
public:
    explicit ComponentDecorator(std::unique_ptr<Component> inner) : inner_(std::move(inner)) {}

protected:
    std::unique_ptr<Component> inner_;
};

class AuditDecorator final : public ComponentDecorator {
public:
    using ComponentDecorator::ComponentDecorator;
    [[nodiscard]] std::string render() const override { return "audit(" + inner_->render() + ")"; }
};

class EncryptDecorator final : public ComponentDecorator {
public:
    using ComponentDecorator::ComponentDecorator;
    [[nodiscard]] std::string render() const override { return "enc(" + inner_->render() + ")"; }
};

int main() {
    PlainMessage base;
    std::cout << "base=" << base.render() << '\n';
    std::cout << "audit=" << AuditDecorator(std::make_unique<PlainMessage>()).render() << '\n';
    std::cout << "encrypted=" << EncryptDecorator(std::make_unique<PlainMessage>()).render() << '\n';
    std::cout << "stacked="
              << AuditDecorator(std::make_unique<EncryptDecorator>(std::make_unique<PlainMessage>())).render()
              << '\n';
}

File: src/Systems/Go/decorator.go
package main

import "fmt"

type component interface {
	render() string
}

type plainMessage struct{}

func (plainMessage) render() string { return "alert" }

type auditDecorator struct{ inner component }

func (d auditDecorator) render() string { return "audit(" + d.inner.render() + ")" }

type encryptDecorator struct{ inner component }

func (d encryptDecorator) render() string { return "enc(" + d.inner.render() + ")" }

func main() {
	var base component = plainMessage{}
	fmt.Println("base=" + base.render())
	fmt.Println("audit=" + (auditDecorator{inner: base}).render())
	fmt.Println("encrypted=" + (encryptDecorator{inner: base}).render())
	fmt.Println("stacked=" + (auditDecorator{inner: encryptDecorator{inner: base}}).render())
}
